phoenix-core/src/main/java/org/apache/phoenix/compile/IndexColumnExpressionCompiler.java

```diff
@@ -0,0 +1,302 @@
+import org.apache.phoenix.parse.ParseNode;
+import org.apache.phoenix.parse.RowValueConstructorParseNode;
+import org.apache.phoenix.parse.SQLParser;
+import org.apache.phoenix.parse.StringConcatParseNode;
+import org.apache.phoenix.parse.SubtractParseNode;
+import org.apache.phoenix.parse.TableName;
+import org.apache.phoenix.schema.PColumn;
+import org.apache.phoenix.schema.PTable;
+import org.apache.phoenix.schema.PTableType;
+import org.apache.phoenix.util.IndexUtil;
+
+/**
+ * Visitor that checks if nodes other than ColumnParseNode are present as 
+ * an expression in an index. If the expression is present, then the node is
+ * not visited but processed as a ColumnParseNode.
+ */
+public class IndexColumnExpressionCompiler extends ExpressionCompiler {
+	
+	IndexColumnExpressionCompiler(StatementContext context) {
+        super(context);
+    }
+
+	IndexColumnExpressionCompiler(StatementContext context, boolean resolveViewConstants) {
+        super(context, resolveViewConstants);
+    }
+
+	IndexColumnExpressionCompiler(StatementContext context, GroupBy groupBy) {
+		super(context, groupBy);
+    }
+
+	IndexColumnExpressionCompiler(StatementContext context, GroupBy groupBy, boolean resolveViewConstants) {
+		super(context, groupBy, resolveViewConstants);
+    }
+	
+    /**
+     * Returns the compiled expression
+     * 
+     * @param node
+     *            data table parse node
+     */
+	private Expression getExpression(ParseNode node) throws SQLException {
+		PhoenixConnection connection = this.context.getConnection();
+        PTable indexTable = this.context.getCurrentTable().getTable();
+        NamedTableNode dataTableNode = NamedTableNode.create(null, TableName.create(indexTable.getParentSchemaName().getString(), 
+        		indexTable.getParentTableName().getString()), Collections.<ColumnDef>emptyList());
+        ColumnResolver resolver = FromCompiler.getResolver(dataTableNode, connection);
+		StatementContext context = new StatementContext(this.context.getStatement(), resolver, new Scan(), new SequenceManager(this.context.getStatement()));        
+        ExpressionCompiler expressionCompiler = new ExpressionCompiler(context);
+        return node.accept(expressionCompiler);		
+	}
+
+	/**
+	 * @return true if current table is an index and there is an expression that matches the given node
+	 */
+	private boolean matchesIndexedExpression(ParseNode node) throws SQLException {   
+		if (context.getCurrentTable().getTable().getType()!=PTableType.INDEX) {
+			return false;
+		}
+		DataTableNodeRewriter statementRewriter = new DataTableNodeRewriter();
+        ParseNode dataTableParseNode = node.accept(statementRewriter);
+		Expression dataTableExpression = getExpression(dataTableParseNode);
+		// if the node matches a column expression
+		if (ColumnExpression.class.isAssignableFrom(dataTableExpression.getClass())) {
+			return false;
+		}
+	    for ( PColumn column : this.context.getCurrentTable().getTable().getColumns()) {
```

Comment:
In IndexColumnExpressionCompiler, all the visitEnter methods check if the ParseNode matches an expression in the index, if there is a match  it returns false for visitEnter. Then in the visitLeave I call convertAndVisitParseNode which creates a ColumnParseNode for the column that contains the matching expression and then calls the visit method on the column parse node. 

Instead of this, should I only check if there is a match in the visitLeave ?